proof/refine/ARM/TcbAcc_R.thy

```diff
@@ -1886,8 +1886,9 @@ lemma asUser_setRegister_corres:
 end
 
 lemma getThreadState_corres:
-  "corres thread_state_relation (tcb_at t) (tcb_at' t)
-          (get_thread_state t) (getThreadState t)"
+  "t = t' \<Longrightarrow>
+   corres thread_state_relation (tcb_at t) (tcb_at' t)
+          (get_thread_state t) (getThreadState t')"
```

Comment:
I am finding it hard to appreciate the merit of this change. It seems odd to talk about different pointers when we assume `obj_at` and `obj_at’` for the same `t`. And it does make to assume those for such a basic function. So, probably it is a bit too much to apply this change for this case.  The same for the `isRunnable_corres` below.

Reply:
or maybe just change the `t`’s in the assumptions as well and it might work, actually. But it is generating more work to apply in many places it seems.

Reply:
These rules will always be applied when the guards are schematic, so it doesn't matter what the guards say (I could also use a cross lemma so that the concrete guard is just `Top` if that is preferred, though there might not be much benefit). 

I was really just wanting to make it easier to apply these sorts of rules (without first having to do a `simp` or a `corres_gen_asm` or something). Do we now think that this is not a good idea to have `corres` rules in this form? Or would Corey's suggestion with the `lemmas` statement be a good compromise?

Reply:
But is it right to say that, with this change, we will “always” have to simplify after applying the rule? Doesn't that mean this gives us more work overall, compared with having to maybe preprocess sometimes? I agree Corey’s suggestion is one solution, and if we are doing that, it probably makes sense to just keep the original statement with the original name so that we don’t need to update the existing proofs, and use the name with prime (‘) for the new version to be used only when it is needed.

Reply:
Yes we will always have to simplify after applying a rule in this form. 

I am not sure whether it would be more work, but it certainly seems more convenient to me to always have the rule apply, and then resolve this obligation with a `simp` or `clarsimp simp: tcb_relation_def` or something after. We discussed this in the RT channel and there were not any objections then. From memory, I have often had to do `corres_req` or `corres_gen_asm` before applying the rule, which is much more manual and prone to breaking. @corlewis @lsf37 do you have any further thoughts on this?

To avoid having to state and prove similar lemmas twice, I think the `lemmas` statement is the best approach. So I could have a primed name for the more complicated version, and then give the `lemmas` statement the unprimed name

Reply:
My understanding is that we often need this sort of assumption to handle cases where the various bits of machinery haven't already worked out that they are the same t's. I think it's a bit simpler here than with similar cases in hoare triple, in that due to the way that we use return value relations we can have this as an assumption and not a precondition. However, my feeling is that at the moment it is as Michael describes, with us needing to have some sort of step either before or after applying the corres rule.

Assuming that we're not going to invest the time to try and find some sort of solution like making `corressimp` more powerful, my preference is something along the lines of what is here. Having the simpler `lemmas` version is a bit of a trade-off, it makes things simpler for us but I can see it being a bit confusing for new proof engineers in the future. Even considering that, I prefer having the simpler version for when it can be used.

What I'm not certain about is the question about which `t` to refer to in the preconditions. Now that I'm looking at it, it does feel a bit odd to have the wrong `t` but also it logically doesn't make a difference and generally we do prefer to keep things in terms of their abstract versions.

Reply:
I don’t think this style of giving the equality will “always” work, though. If you need the state relation, or even some of the preconditions on either of the abs/conc sides (such as `valid_objs`), to show the equality, you will get stuck, unless you have shown the equality using the usual `corres_req`, etc., prior to applying the rule. It might be rare to have such a situation but I don’t think we can exclude the possibility, particularly with MCS. I suppose we can backtrack if we get stuck like that and show what needs to be shown, which I guess could still work overall.